### ISSUE TITLE
Error in 0.0.23 When Using Git Annotated Tags

### ISSUE THREAD

**Issue by hartmantis** [applicationsonline/librarian#76]
I recently installed Librarian on a new machine and it pulled down the new 0.0.23. I've noticed that, in this version, it seems to error out when I have a cookbook in Git with a ref that's an annotated tag. Example Cheffile:

```
#!/usr/bin/env ruby
#^syntax detection

site 'http://community.opscode.com/api/v1'

cookbook 'just-a-test-cookbook',
    :git => 'git://github.com/RoboticCheese/just-a-test-cookbook.git',
    :ref => 'annotated_tag'

#cookbook 'just-a-test-cookbook',
#    :git => 'git://github.com/RoboticCheese/just-a-test-cookbook.git',
#    :ref => 'lightweight_tag'
```

The `lightweight_tag` ref commented out above checks out fine if it's uncommented, but `annotated_tag` gives this verbose output:

```
[Librarian] Ruby Version: 1.8.7
[Librarian] Ruby Platform: universal-darwin11.0
[Librarian] Rubygems Version: 1.8.24
[Librarian] Librarian Version: 0.0.23
[Librarian] Librarian Adapter: chef
[Librarian] Project: /Users/jhartman/Desktop/testing-librarian
[Librarian] Specfile: Cheffile
[Librarian] Lockfile: Cheffile.lock
[Librarian] Git: /usr/bin/git
[Librarian] Git Version: git version 1.7.7.5 (Apple Git-26)
[Librarian] Git Environment Variables:
[Librarian]   (empty)
[Librarian] Pre-Cached Sources:
[Librarian] Post-Cached Sources:
[Librarian]   [:git, "git://github.com/RoboticCheese/just-a-test-cookbook.git", {:ref=>"annotated_tag"}]
[Librarian]   [:site, "http://community.opscode.com/api/v1", {}] 
[Librarian] Scheduling just-a-test-cookbook (>= 0) <git://github.com/RoboticCheese/just-a-test-cookbook.git#annotated_tag>
[Librarian] Resolving just-a-test-cookbook (>= 0) <git://github.com/RoboticCheese/just-a-test-cookbook.git#annotated_tag>
[Librarian]   No known prior constraints
[Librarian]   Checking manifests
[Librarian] Running `/usr/bin/git clone git://github.com/RoboticCheese/just-a-test-cookbook.git . --quiet` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     --- No output
[Librarian] Running `/usr/bin/git reset --hard --quiet` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     --- No output
[Librarian] Running `/usr/bin/git clean -x -d --force --force` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     --- No output
[Librarian] Running `/usr/bin/git rev-parse HEAD --quiet` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     --> b9e0c49ebf01757684263eade300a232d0d4bdc9
[Librarian] Running `/usr/bin/git fetch origin --quiet` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     --- No output
[Librarian] Running `/usr/bin/git fetch origin --quiet --tags` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     --- No output
[Librarian] Running `/usr/bin/git remote` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     --> origin
[Librarian] Running `/usr/bin/git branch -r` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     -->   origin/HEAD -> origin/master
[Librarian]     -->   origin/master
[Librarian] Running `/usr/bin/git rev-parse annotated_tag --quiet` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     --> 5e63b02a367a486cc28d18ab60220ac3d45617eb
[Librarian] Running `/usr/bin/git rev-parse HEAD --quiet` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     --> b9e0c49ebf01757684263eade300a232d0d4bdc9
[Librarian] Running `/usr/bin/git checkout 5e63b02a367a486cc28d18ab60220ac3d45617eb --quiet` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     --- No output
[Librarian] Running `/usr/bin/git rev-parse HEAD --quiet` in tmp/librarian/cache/source/git/3b9defe12d10aae498945d8735cd22b0
[Librarian]     --> b9e0c49ebf01757684263eade300a232d0d4bdc9
failed to checkout 5e63b02a367a486cc28d18ab60220ac3d45617eb
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/source/git.rb:109:in `cache!'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/source/local.rb:31:in `manifest_search_paths'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/source/local.rb:40:in `found_path'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/chef/source/local.rb:54:in `expect_manifest!'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/chef/source/local.rb:42:in `fetch_manifest_data'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/chef/source/local.rb:38:in `manifest_data'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/chef/source/local.rb:27:in `fetch_version'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/manifest.rb:117:in `fetch_version!'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/manifest.rb:109:in `fetched_version'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/manifest.rb:63:in `version'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/manifest.rb:59:in `to_s'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver/implementation.rb:67:in `recursive_resolve'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver/implementation.rb:113:in `debug'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/helpers/debug.rb:28:in `debug'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/ui.rb:32:in `debug'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/helpers/debug.rb:28:in `debug'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver/implementation.rb:113:in `debug'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver/implementation.rb:67:in `recursive_resolve'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver/implementation.rb:64:in `each'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver/implementation.rb:64:in `recursive_resolve'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver/implementation.rb:107:in `scope'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver/implementation.rb:63:in `recursive_resolve'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver/implementation.rb:107:in `scope'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver/implementation.rb:46:in `recursive_resolve'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver/implementation.rb:24:in `resolve'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/resolver.rb:19:in `resolve'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/action/resolve.rb:25:in `run'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/cli.rb:139:in `resolve!'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/cli.rb:75:in `install'
/Library/Ruby/Gems/1.8/gems/thor-0.15.0/lib/thor/task.rb:27:in `send'
/Library/Ruby/Gems/1.8/gems/thor-0.15.0/lib/thor/task.rb:27:in `run'
/Library/Ruby/Gems/1.8/gems/thor-0.15.0/lib/thor/invocation.rb:120:in `invoke_task'
/Library/Ruby/Gems/1.8/gems/thor-0.15.0/lib/thor.rb:275:in `dispatch'
/Library/Ruby/Gems/1.8/gems/thor-0.15.0/lib/thor/base.rb:408:in `start'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/lib/librarian/cli.rb:33:in `bin!'
/Library/Ruby/Gems/1.8/gems/librarian-0.0.23/bin/librarian-chef:7
/usr/bin/librarian-chef:23:in `load'
/usr/bin/librarian-chef:23
```

This worked previously in 0.0.22. I've tried it in 0.0.23 and built from master on a few machines with different OSes and versions of Ruby/Git and had the same result. Looked at the source briefly in hopes of contributing, but a bit of a Ruby noob over here.

Thanks!

**Comment by yfeldblum** [comment]
In 0.0.23, the git source starts verifying that the expected SHA is what actually got checked out by doing the following (effectively):

```
expected_sha = `git rev-parse #{ref}`
actual_sha = `git rev-parse HEAD`
assert_equal expected_sha, actual_sha
```

With the name of an annotated tag for the ref, however, the SHA that gets checked out is not the same as the annotated tag's SHA. So there needs to be some way to handle this case reliably.

**Comment by yfeldblum** [comment]
Does d97c12d fix it for you?

**Comment by hartmantis** [comment]
Looks like it does. Yay, thank you!

**Comment by yfeldblum** [comment]
Fixed by d97c12d.
